### PR TITLE
Feature/gh-308: Add Type Safety to UserModule

### PR DESCRIPTION
**Pull request by MitchellScottSandre**

### CODE REVIEW

File: projects/storefrontlib/src/lib/user/store/effects/user-addresses.effect.ts [hunk 1]
```diff
@@ -16,7 +16,7 @@ export class UserAddressesEffects {
       return this.occUserService.loadUserAddresses(payload).pipe(
         map((addressesList: any) => {
           return new fromUserAddressesAction.LoadUserAddressesSuccess(
-            addressesList.addresses
+            addressesList
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/user/store/effects/user-addresses.effect.ts]: the response is this:
```
{
   "addresses" : [ {
      "country" : {
         "isocode" : "CA"
      },
      "defaultAddress" : true,
      ...
   }, {
      "country" : {
         "isocode" : "CA"
      },
      "defaultAddress" : false,
      ...
   } ]
}
```
So,to get the array [], we need `addressList.address`

File: projects/storefrontlib/src/lib/user/store/actions/order-details.action.spec.ts [hunk 2]
```diff
@@ -1,17 +1,27 @@
 import * as fromOrderDetailsAction from './order-details.action';
+import { Order } from '@spartacus/core';
 
-const mockUserOrder: any = {
-  orderId: '123'
+// const mockUserOrder: any = {
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/user/store/actions/order-details.action.spec.ts]: remove comment

File: projects/storefrontlib/src/lib/user/store/actions/payment-methods.action.spec.ts [hunk 3]
```diff
@@ -31,7 +32,9 @@ describe('User Payment Methods Actions', () => {
   });
 
   describe('LoadUserPaymentMethodsSuccess Action', () => {
-    const mockUserPaymentMethods = { payments: ['payment1', 'payment2'] };
+    const mockUserPaymentMethods: PaymentDetailsList = {
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/user/store/actions/payment-methods.action.spec.ts]: pass `PaymentDetails[]` to `LoadUserPaymentMethodsSuccess`

File: projects/storefrontlib/src/lib/user/store/actions/payment-methods.action.ts [hunk 4]
```diff
@@ -18,7 +19,7 @@ export class LoadUserPaymentMethodsFail implements Action {
 
 export class LoadUserPaymentMethodsSuccess implements Action {
   readonly type = LOAD_USER_PAYMENT_METHODS_SUCCESS;
-  constructor(public payload: any) {}
+  constructor(public payload: PaymentDetailsList) {}
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/user/store/actions/payment-methods.action.ts]: `paymentDetails[]`.

File: projects/storefrontlib/src/lib/user/store/effects/payment-methods.effect.ts [hunk 5]
```diff
@@ -17,9 +18,9 @@ export class UserPaymentMethodsEffects {
     ),
     mergeMap(payload => {
       return this.occUserService.loadUserPaymentMethods(payload).pipe(
-        map((paymentsList: any) => {
+        map((paymentsList: PaymentDetailsList) => {
           return new fromUserPaymentMethodsAction.LoadUserPaymentMethodsSuccess(
-            paymentsList.payments
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/user/store/effects/payment-methods.effect.ts]: pass `PaymentDetails[]` to `LoadUserPaymentMethodsSuccess`.

File: projects/storefrontlib/src/lib/user/store/reducers/payment-methods.reducer.spec.ts [hunk 6]
```diff
@@ -14,7 +15,9 @@ describe('User Payment Methods Reducer', () => {
 
   describe('LOAD_USER_PAYMENT_METHODS_SUCCESS action', () => {
     it('should populate the user Payment Methods state entities', () => {
-      const mockUserPaymentMethods = ['payment1', 'payment2'];
+      const mockUserPaymentMethods: PaymentDetailsList = {
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/user/store/reducers/payment-methods.reducer.spec.ts]: `mockUserPaymentMethods` should be an array `PaymentDetails[]`

File: projects/storefrontlib/src/lib/user/store/reducers/payment-methods.reducer.ts [hunk 7]
```diff
@@ -1,12 +1,13 @@
 import * as fromPaymentMethodsAction from '../actions/payment-methods.action';
+import { PaymentDetailsList } from '@spartacus/core';
 
 export interface UserPaymentMethodsState {
-  list: any;
+  list: PaymentDetailsList;
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/user/store/reducers/payment-methods.reducer.ts]: `PaymentDetails[]`

File: projects/storefrontlib/src/lib/user/store/effects/user-addresses.effect.ts [hunk 8]
```diff
@@ -16,7 +16,7 @@ export class UserAddressesEffects {
       return this.occUserService.loadUserAddresses(payload).pipe(
         map((addressesList: any) => {
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/user/store/effects/user-addresses.effect.ts]: change `any` to `AddressList`

File: projects/storefrontlib/src/lib/user/store/selectors/delivery-countries.selectors.ts [hunk 9]
```diff
@@ -17,7 +17,7 @@ export const getDeliveryCountriesEntites: MemoizedSelector<
 
 export const getAllDeliveryCountries: MemoizedSelector<
   any,
-  any
+  any[]
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/user/store/selectors/delivery-countries.selectors.ts]: Can we use type `Country` here?

File: projects/storefrontlib/src/lib/user/store/selectors/payment-methods.selectors.ts [hunk 10]
```diff
@@ -12,7 +13,10 @@ export const getPaymentMethodsState: MemoizedSelector<
   (state: fromFeature.UserState) => state.payments
 );
 
-export const getPaymentMethods: MemoizedSelector<any, any> = createSelector(
+export const getPaymentMethods: MemoizedSelector<
+  any,
+  PaymentDetailsList
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/user/store/selectors/payment-methods.selectors.ts]: `PaymentDetails[]`

File: projects/storefrontlib/src/lib/user/store/selectors/titles.selectors.ts [hunk 11]
```diff
@@ -7,12 +7,12 @@ export const getTitlesState = createSelector(
   (state: fromFeature.UserState) => state.titles
 );
 
-export const getTitlesEntites: MemoizedSelector<any, any> = createSelector(
-  getTitlesState,
-  fromReducer.getTitlesEntites
-);
+export const getTitlesEntites: MemoizedSelector<
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/user/store/selectors/titles.selectors.ts]: Can we use type `Title` in this selector?

File: projects/storefrontlib/src/lib/user/store/selectors/user-addresses.selectors.spec.ts [hunk 12]
```diff
@@ -1,11 +1,11 @@
 import { TestBed } from '@angular/core/testing';
 import { Store, StoreModule, select } from '@ngrx/store';
-
+import { Address } from '@spartacus/core';
 import * as fromActions from '../actions';
 import * as fromReducers from '../reducers';
 import * as fromSelectors from '../selectors';
 
-const mockUserAddresses = ['address1', 'address2'];
```
Comment (WeizhengSap) [inline projects/storefrontlib/src/lib/user/store/selectors/user-addresses.selectors.spec.ts]: select an array, not object.

**Comment by MitchellScottSandre** [comment]
fixes GH-308

**Pull request merged**